File: parse_quiz.js
const fs = require('fs');
const path = require('path');

const TXT_DIR = 'E:/wenjian/frist cc/mianyi_txt';
const OUT_JSON = 'E:/wenjian/frist cc/quiz-data.json';

function parseChapterNum(filename) {
  const m = filename.match(/^(\d+)/);
  return m ? parseInt(m[1]) : 999;
}

function splitOptionsInLine(line) {
  const normalized = line.replace(/\s*([A-E])\s*[．.、\s]\s*/g, '\n$1. ');
  return normalized.split('\n')
    .map(s => s.trim())
    .filter(s => /^[A-E]\.\s/.test(s));
}

// ── 通用题目解析 ──
function parseQuestionBlock(text, qLabel, endLabels) {
  const idx = text.indexOf(qLabel);
  if (idx < 0) return [];

  const section = text.slice(idx);
  let end = section.length;
  for (const m of endLabels) {
    const p = section.indexOf(m);
    if (p > 0 && p < end) end = p;
  }

  const block = section.slice(0, end);
  const parts = block.split(/\n\s*(\d+)\s*[．.、]\s*/);
  const questions = [];

  for (let i = 1; i + 1 < parts.length; i += 2) {
    const qnum = parseInt(parts[i]);
    const content = parts[i + 1];
    const lines = content.split('\n');

    let title = '';
    const options = [];

    for (const rawLine of lines) {
      let line = rawLine.trim();
      if (!line) continue;

      if (/^[A-E]\s*[．.、\s]/.test(line)) {
        const splitOpts = splitOptionsInLine(line);
        for (const opt of splitOpts) {
          if (options.length < 5) options.push(opt);
        }
      } else if (options.length === 0) {
        title += line;
      } else if (/[A-E]\s*[．.、\s]/.test(line)) {
        const splitOpts = splitOptionsInLine(line);
        for (const opt of splitOpts) {
          if (options.length < 5) options.push(opt);
        }
      } else {
        if (options.length > 0) {
          options[options.length - 1] += ' ' + line;
        }
      }
    }

    if (title.trim() && options.length >= 3) {
      questions.push({ id: qnum, title: title.trim(), options });
    }
  }
  return questions;
}

// ── A 型答案解析（单选） ──
function parseAAnswersFromBlock(ansBlock) {
  const pairs = [...ansBlock.matchAll(/(\d+)\s*[．.\s、，,]*\s*([A-E])(?![A-E])/g)];
  const answers = {};
  for (const [, numStr, letter] of pairs) {
    answers[parseInt(numStr)] = letter.charCodeAt(0) - 'A'.charCodeAt(0);
  }
  return answers;
}

// ── X 型答案解析（多选） ──
function parseXAnswersFromBlock(ansBlock) {
  // Match patterns like: "1 ABCDE" or "1.ABCDE" or "1．ABCDE" or "1 AB, 2 BC"
  const answers = {};
  // match "number" followed by "multiple letters"
  const re = /(\d+)\s*[．.\s、，:：]*\s*([A-E]{2,5})/gi;
  let match;
  while ((match = re.exec(ansBlock)) !== null) {
    const num = parseInt(match[1]);
    const letters = match[2].toUpperCase().split('');
    const indices = letters.map(l => l.charCodeAt(0) - 'A'.charCodeAt(0));
    if (!answers[num]) answers[num] = [];
    answers[num] = [...new Set([...answers[num], ...indices])].sort();
  }
  return answers;
}

// ── 查找答案区块 ──
function findAnswerSection(text) {
  for (const m of ['参考答案', '题解']) {
    const p = text.indexOf(m);
    if (p >= 0) return text.slice(p);
  }
  return null;
}

function findAnswerBlock(ansText, label, altLabel, endLabels) {
  let idx = ansText.indexOf(label);
  if (idx < 0 && altLabel) idx = ansText.indexOf(altLabel);
  if (idx < 0) return null;

  let block = ansText.slice(idx);
  let end = block.length;
  for (const m of endLabels) {
    const p = block.indexOf(m);
    if (p > 0 && p < end) end = p;
  }
  return block.slice(0, end);
}

// ── 解析单个文件 ──
function parseFile(filepath) {
  const text = fs.readFileSync(filepath, 'utf-8');
  const lines = text.split('\n').map(l => l.trim()).filter(Boolean);

  let title = lines[0] || path.basename(filepath);
  title = title.replace(/^第[一二三四五六七八九十\d]+章\s*/, '').trim();
  if (!title) title = lines[1] || '未知章节';

  // 题目
  const allEndLabels = ['[B型题]', '[B 型题]', '[C型题]', '[C 型题]', '[X型题]', '[X 型题]',
    '三、', '四、', '五、', '一、'];
  const aquestions = parseQuestionBlock(text, '[A型题]', allEndLabels);
  const xquestions = parseQuestionBlock(text, '[X型题]', ['[B型题]', '[B 型题]', '[C型题]', '[C 型题]',
    '三、', '四、', '五、', '一、', '二、']);

  // 答案
  const ansText = findAnswerSection(text);
  const aanswers = {};
  const xanswers = {};
  if (ansText) {
    const ansEndLabels = ['[B型题]', '[B 型题]', '[C型题]', 'B型题：', 'C型题：',
      '三、', '四、', '五、', '二、'];
    const aBlock = findAnswerBlock(ansText, '[A型题]', 'A型题：', ansEndLabels);
    if (aBlock) Object.assign(aanswers, parseAAnswersFromBlock(aBlock));

    const xEndLabels = ['[B型题]', '[B 型题]', '[C型题]', 'B型题：', 'C型题：',
      '三、', '四、', '五、', '二、', '一、'];
    const xBlock = findAnswerBlock(ansText, '[X型题]', 'X型题：', xEndLabels);
    if (xBlock) {
      const xParsed = parseXAnswersFromBlock(xBlock);
      // If nothing found after [X型题], also look before it (some docs put X answers before the label)
      if (Object.keys(xParsed).length === 0) {
        const xIdx = ansText.indexOf('[X型题]');
        if (xIdx > 0) {
          // Search up to 500 chars before [X型题] for X-type answer patterns
          const preBlock = ansText.slice(Math.max(0, xIdx - 500), xIdx);
          Object.assign(xParsed, parseXAnswersFromBlock(preBlock));
        }
      }
      Object.assign(xanswers, xParsed);
    }
  }

  for (const q of aquestions) {
    q.answer = aanswers[q.id] !== undefined ? aanswers[q.id] : -1;
  }
  for (const q of xquestions) {
    q.answer = xanswers[q.id] !== undefined ? xanswers[q.id] : [];
  }

  return { title, aquestions, xquestions };
}

// ── 主程序 ──
function main() {
  const files = fs.readdirSync(TXT_DIR)
    .filter(f => f.endsWith('.txt'))
    .map(f => ({ name: f, num: parseChapterNum(f) }))
    .sort((a, b) => a.num - b.num);

  const chapters = [];
  let totalA = 0, totalX = 0;
  const issues = [];

  for (const { name, num } of files) {
    const data = parseFile(path.join(TXT_DIR, name));
    const aWithAns = data.aquestions.filter(q => q.answer >= 0).length;
    const xWithAns = data.xquestions.filter(q => q.answer.length > 0).length;

    console.log(`第${String(num).padStart(2)}章 ${data.title}: A型${data.aquestions.length}题(${aWithAns}有答案) X型${data.xquestions.length}题(${xWithAns}有答案)`);

    if (data.aquestions.length === 0 && data.xquestions.length === 0) {
      issues.push(`第${num}章: 未解析到任何题目`);
    }
    const aMissing = data.aquestions.filter(q => q.answer < 0).map(q => q.id);
    if (aMissing.length > 0) issues.push(`第${num}章 A型: 题目 ${aMissing.join(', ')} 缺少答案`);
    const xMissing = data.xquestions.filter(q => q.answer.length === 0).map(q => q.id);
    if (xMissing.length > 0) issues.push(`第${num}章 X型: 题目 ${xMissing.join(', ')} 缺少答案`);

    chapters.push({ num, title: data.title, filename: name, aquestions: data.aquestions, xquestions: data.xquestions });
    totalA += data.aquestions.length;
    totalX += data.xquestions.length;
  }

  fs.writeFileSync(OUT_JSON, JSON.stringify(chapters, null, 2), 'utf-8');

  console.log(`\n总计: ${chapters.length} 章, A型 ${totalA} 题, X型 ${totalX} 题, 共 ${totalA + totalX} 题`);
  if (issues.length > 0) {
    console.log('\n⚠️ 问题:');
    issues.forEach(i => console.log('  - ' + i));
  }
}

main();

File: parse_quiz.py
import re, json, os, glob

TXT_DIR = r"E:\wenjian\frist cc\mianyi_txt"
OUT_JSON = r"E:\wenjian\frist cc\quiz-data.json"

def parse_chapter_num(filename):
    m = re.match(r'(\d+)', filename)
    if m: return int(m.group(1))
    # Handle "11、12" case
    m = re.match(r'(\d+)[、,]', filename)
    return int(m.group(1)) if m else 999

def parse_questions(text):
    """Find A-type questions in the QUESTIONS section (first [A型题])"""
    # Find first [A型题] - this is the questions section
    idx = text.find('[A型题]')
    if idx < 0:
        idx = text.find('[A 型题]')
    if idx < 0:
        return []

    # Get text from [A型题] to the next [B型题] or [C型题] or [X型题] or end of file
    section = text[idx:]
    end_markers = ['[B型题]', '[B 型题]', '[C型题]', '[C 型题]', '[X型题]', '[X 型题]']
    end = len(section)
    for marker in end_markers:
        pos = section.find(marker)
        if pos > 0 and pos < end:
            end = pos
    section = section[:end]

    questions = []
    # Split by question number pattern: number followed by ．or . or 、
    # Match patterns like "1．", "2.", "3、", "1 ." etc
    parts = re.split(r'\n\s*(\d+)\s*[．.、]\s*', '\n' + section)
    # parts[0] = before first question, parts[1] = num1, parts[2] = content1, parts[3] = num2, ...

    i = 1
    while i + 1 < len(parts):
        qnum = parts[i]
        content = parts[i + 1]
        i += 2

        # Parse title and options
        title = ''
        options = []
        lines = content.strip().split('\n')

        for line in lines:
            line = line.strip()
            if not line:
                continue
            # Check if line starts with an option letter (A-E) followed by ．or . or 、or space
            opt_match = re.match(r'^([A-E])\s*[．.、\s]\s*(.*)', line)
            if opt_match:
                options.append(opt_match.group(1) + '. ' + opt_match.group(2).strip())
            elif not options:
                # This is part of the title (before any option found)
                title += line
            else:
                # Could be continuation of last option text
                if options:
                    options[-1] += ' ' + line

        if title and len(options) == 5:
            # Clean title
            title = title.strip()
            questions.append({
                'id': int(qnum),
                'title': title,
                'options': options
            })
        elif title and len(options) > 0:
            # Some questions might have fewer/more options - include anyway
            questions.append({
                'id': int(qnum),
                'title': title.strip(),
                'options': options
            })

    return questions

def parse_answers(text):
    """Parse A-type answers from the answer section"""
    # Find answer section
    ans_section_start = -1
    for marker in ['参考答案', '参考 答案']:
        idx = text.find(marker)
        if idx >= 0:
            ans_section_start = idx
            break

    if ans_section_start < 0:
        return {}

    ans_text = text[ans_section_start:]

    # Find [A型题] or "A型题：" in answer section
    idx = ans_text.find('[A型题]')
    if idx < 0:
        idx = ans_text.find('[A 型题]')
    if idx < 0:
        idx = ans_text.find('A型题：')
    if idx < 0:
        idx = ans_text.find('A型题:')
    if idx < 0:
        return {}

    # Get the answer line(s) - from marker to the next section marker or end
    ans_block = ans_text[idx:]
    end_markers = ['[B型题]', '[B 型题]', '[C型题]', 'B型题：', 'C型题：',
                   '二、', '三、', '四、', '五、']
    end = len(ans_block)
    for marker in end_markers:
        pos = ans_block.find(marker)
        if pos > 0 and pos < end:
            end = pos
    ans_block = ans_block[:end]

    # Parse all "number letter" pairs
    # Match patterns like: 1．D, 2.B, 3 E, 1 D，2 B, 4.C, etc.
    pairs = re.findall(r'(\d+)\s*[．.、\s]*\s*([A-E])', ans_block)

    answers = {}
    for num_str, letter in pairs:
        num = int(num_str)
        answers[num] = ord(letter) - ord('A')

    return answers

def parse_file(filepath):
    with open(filepath, 'r', encoding='utf-8', errors='ignore') as f:
        text = f.read()

    # Get chapter title from first non-empty line
    lines = [l.strip() for l in text.split('\n') if l.strip()]
    title = lines[0] if lines else os.path.basename(filepath)
    # Clean title: remove leading numbers like "第一章  " or "第二章      "
    title = re.sub(r'^第[一二三四五六七八九十\d]+章\s*', '', title).strip()
    if not title:
        title = lines[1] if len(lines) > 1 else '未知章节'

    questions = parse_questions(text)
    answers = parse_answers(text)

    # Merge answers into questions
    for q in questions:
        if q['id'] in answers:
            q['answer'] = answers[q['id']]
        else:
            q['answer'] = -1  # unknown

    return {
        'title': title,
        'questions': questions
    }

def main():
    files = sorted(glob.glob(os.path.join(TXT_DIR, '*.txt')),
                   key=lambda f: parse_chapter_num(os.path.basename(f)))

    chapters = []
    total_questions = 0
    issues = []

    for f in files:
        fname = os.path.basename(f)
        ch_num = parse_chapter_num(fname)
        data = parse_file(f)

        q_with_answers = sum(1 for q in data['questions'] if q['answer'] >= 0)
        q_total = len(data['questions'])

        print(f"第{ch_num:2d}章 {data['title']}: {q_total} 题, {q_with_answers} 有答案")

        if q_total == 0:
            issues.append(f"第{ch_num}章: 未解析到题目")
        if q_with_answers < q_total:
            missing = [str(q['id']) for q in data['questions'] if q['answer'] < 0]
            issues.append(f"第{ch_num}章: 题目 {', '.join(missing)} 缺少答案")

        chapters.append({
            'num': ch_num,
            'title': data['title'],
            'filename': fname,
            'questions': data['questions']
        })
        total_questions += q_total

    with open(OUT_JSON, 'w', encoding='utf-8') as f:
        json.dump(chapters, f, ensure_ascii=False, indent=2)

    print(f"\n总计: {len(chapters)} 章, {total_questions} 题")
    if issues:
        print("\n⚠️ 问题:")
        for iss in issues:
            print(f"  - {iss}")

if __name__ == '__main__':
    main()
